lib/AssemblyUtil/FastaToAssembly.py

```diff
@@ -100,15 +139,39 @@ def _import_fasta_mass(self, params):
                 json.dump(ao, f)
 
         # save to WS and return
-        assembly_infos = self._save_assembly_objects(
-            params[_WSID],
-            [p[_ASSEMBLY_NAME] for p in params[_INPUTS]],
-            assobjects
-        )
+        assembly_infos = []
+        assembly_names = [p[_ASSEMBLY_NAME] for p in params[_INPUTS]]
+        # generator to process files in bach. Avoid memory issues
+        for assobject_batch, assembly_name_batch in self._assembly_objects_generator(
+            assobjects, assembly_names, max_cumsize
+        ):
+            assembly_infos.extend(
+                self._save_assembly_objects(params[_WSID], assembly_name_batch, assobject_batch)
+            )
+
         for out, ai in zip(output, assembly_infos):
             out['upa'] = _upa(ai)
         return output
 
+    def _run_parallel_import_fasta_mass(self, params, workers):
+        print(f' - running {workers} parallel workers')
+
+        # distribute inputs evenly across workers
+        param_inputs = params.pop(_INPUTS)
+        chunk_size = math.ceil(len(param_inputs) / workers)
+        batch_input = [
+            (
+                {
+                    **params,
+                    _INPUTS: param_inputs[i : i + chunk_size]
+                }
+            )
+            for i in range(0, len(param_inputs), chunk_size)
+        ]
+        batch_result = Pool(workers).map(self._import_fasta_mass, batch_input)
+        result = list(itertools.chain.from_iterable(batch_result))
```

Comment:
What happens here if one of the workers threw an error?

Reply:
If my understanding is correct, that means save_assemblies_from_fastas will throw an error as well, which will catch by except clause [here](https://github.com/kbase/collections/blob/66bb7997e4a0225f55c7b17ba6e62892d78fe0ce/src/loaders/workspace_uploader/workspace_uploader.py#L381). And we need to check workspace to rm potential partial upload.

Reply:
Can you test that your understanding is correct, ideally in a test but at least manually?

Reply:
I kinda tested it before my https://github.com/kbaseapps/AssemblyUtil/pull/89#discussion_r1358992625. (I passed an array of max_cumsize [1,2,3] and let the worker failed if max_cumsize==2). I am curious is there a better way to intentionally let a single worker fail in a test setting.

Reply:
This is kinda tricky because _validate_mass_params step takes place before multiprocessing step.

Reply:
All that does is check that the string is truthy. The workspace is much more stringent than that

Reply:
Got Java Exception: illegal character in object name and then hung. Not quite sure why cannot catch it :confused:.

Reply:
That'll be wrapped in a baseclient.ServerException you should be able to catch

Reply:
It hung so I cancelled the workflow. The stack trace can be found [here](https://github.com/kbaseapps/AssemblyUtil/actions/runs/6551182630/job/17791843148#step:6:1780)

Reply:
Ok, the callback server gets the error. Next step I would do is put some print statements in the DFU client to see what happens when the call to save objects returns

Reply:
It did not return anything and resulted in the same error. 

What I find conflicting is that DataFileUtil should have caught WorkspaceError [here](https://github.com/kbaseapps/DataFileUtil/blob/ee7670582db65adee442f052a52444a0d84a52a0/lib/DataFileUtil/implementation.py#L82) as the stack trace did print this line of code.

Reply:
> It did not return anything and resulted in the same error.

Ok, it that's true that means the callbackserver gets the error from DFU and then never returns from the call from AssemblyUtil. Next step is to dig into what's going on in the callback server and figure out why it's not returning the error to AsssemblyUtil

> What I find conflicting is that DataFileUtil must have caught WorkspaceError [here](https://github.com/kbaseapps/DataFileUtil/blob/ee7670582db65adee442f052a52444a0d84a52a0/lib/DataFileUtil/implementation.py#L82) as the stack trace did print this line of code.

It'll catch it, log it, and reraise it

Reply:
https://kbase.slack.com/archives/C03FK9RKSBX/p1698439559496819?thread_ts=1698341441.490029&cid=C03FK9RKSBX

Updated!

Reply:
:+1: